### ISSUE TITLE
Integrated Language Server Protocol

### ISSUE THREAD

**Issue by schrieveslaach** [nushell/nushell#10794]
### Related problem

As described in https://github.com/nushell/vscode-nushell-lang/issues/117 there is a demand to integrate support for the language server protocol (LSP) into Nushell itself instead of creating an additional command line interface (CLI) protocol (that is the current approach with `--ide-*` CLI options.

### Describe the solution you'd like

Based on the initial draft PR #10723 at least following LSP methods should be supported by Nushell:

- [ ] [textDocument/hover](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_hover)
- [ ] [textDocument/completion](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_completion)
- [ ] [textDocument/definition](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_definition)
- [ ] [textDocument/didChange](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_didChange),  [textDocument/didClose](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_didClose),   and [textDocument/didOpen](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_didOpen)
- [ ] [textDocument/inlayHint](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_inlayHint)
- [ ] [textDocument/publishDiagnostics](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_publishDiagnostics) 
- [ ] [workspace/configuration](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_configuration)
- [ ] [workspace/didChangeConfiguration](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_didChangeConfiguration)

### Describe alternatives you've considered

The alternative would be to keep using the `--ide-*` in the long run and build an external LSP server.

**Comment by fdncred** [comment]
The one other thing that I'd like to see added, which I _think_ has to be added on the vscode side is semantic highlighting. Just logging it here to remember.

**Comment by amtoine** [comment]
just to be sure, will #10723 close this issue? :yum:

**Comment by fdncred** [comment]
> just to be sure, will #10723 close this issue? 😋

Not necessarily. This is the tracking issue. We could land #10723 without everything being completed and then have follow-up PRs to complete the remaining tasks.

**Comment by schrieveslaach** [comment]
@fdncred, `textDocument/semanticTokens` is now in the list. I put an asterisks at the end because there are multiple requests.

**Comment by fdncred** [comment]
> @fdncred, `textDocument/semanticTokens` is now in the list. I put an asterisks at the end because there are multiple requests.

I'm excited about this!!!!

**Comment by fdncred** [comment]
@schrieveslaach When you get a moment, can you update the tracking items above?

**Comment by fdncred** [comment]
@schrieveslaach I just heard that the work you've done here has already been added to `helix` and to `zed`. Here's a screenshot of `zed` that was posted on discord.
![image](https://github.com/nushell/nushell/assets/343840/a62d9375-d2b8-4444-8335-b9c5ab2d5b05)

**Comment by AucaCoyan** [comment]
Hi! I did a lot of work on `nufmt`. I checked out the code for the LSP but sadly I understand very little. I do want to help! LSP give life to my everyday code. Is there any help I can give for you? (particulary for @schrieveslaach ) ❤️

**Comment by schrieveslaach** [comment]
@AucaCoyan, thanks for offering help. In the recent weeks hadn't had much time for working on LSP in Nushell and I wasn't sure what to tackle next. Maybe, on `workspace/configuration` but I'm uncertain how a workspace should behave for a Nushell project.

But what do you think about integrating `nufmt` as a [Document Formatting Request](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_formatting) into Nushell? When Nushell's LSP receives the request, it could look for `nufmt` binary in the PATH variable and then it could format the content kept in memory and sends back the the text edits that the client has to do as well.

**Comment by AucaCoyan** [comment]
Great!
Just for confimation: do you see a better approach to have a different binary called `nufmt` or do you prefer having a library in `nu-lsp` that formats code?
`nufmt` it's not close to be ready as a beta yet, it doesn't format a thing. Just checking if you want this or you prefer having everything in one repo for the lsp 👍🏼

**Comment by fdncred** [comment]
we should not integrate `nufmt` yet. it's not very useful in its current state.

**Comment by fdncred** [comment]
@schrieveslaach one thing I think we need to be able to have is the `-I` include dirs functionality. For vscode, I think that will be a configuration point, but it could also just utilize whatever is in your env.nu NU_LIB_DIRS folders as a starting point.

**Comment by schrieveslaach** [comment]
@fdncred, I'm not so sure if I can follow you. The `-I` option is [evaluated and set put into the `engine_state`](https://github.com/nushell/nushell/blob/1c5bd21ebcdcb86b640af41c31d9d049d7493411/src/main.rs#L242-L251) before [starting the LSP](https://github.com/nushell/nushell/blob/1c5bd21ebcdcb86b640af41c31d9d049d7493411/src/main.rs#L292) (in the same fashion as for the IDE command line options). If you are talking about this, I'm lost what is not yet working.

Or are you talking about the [`rootUri` LSP init parameter](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#initialize)? Here I'm unsure how a `rootUri` should be treated be nu's LSP.

**Comment by fdncred** [comment]
@schrieveslaach Sorry, this is what I'm talking about. It may not be exactly as I've stated but this is the symptom.

This is my config.nu loaded in nvim. Note that the lsp is failing to find some files.
![image](https://github.com/nushell/nushell/assets/343840/6b4cd82d-5006-4c60-9643-1f209fca78ba)

The reason for this is
1. I am not specifying any -I parameter to the nu lsp invocation. I'm kind of a nvim noob and I barely got the lsp working.
2. I'm thinking that the lsp should probably look at NU_LIB_DIRS and apply that automatically. If this were to happen, then I wouldn't have any errors when looking at my config.nu.

In the vscode extension, I get around this by having a config point to manually add Include Dirs.
![image](https://github.com/nushell/nushell/assets/343840/f3945bc0-86b4-4fe2-9613-dad58573e8dc)

I'm not saying we should have a config or not. I'm not sure but I am saying that whatever is in NU_LIB_DIRS should be used. This may also assume that a persons env.nu is parsed when launching with `nu --lsp` which I'm not confident of.

**Comment by schrieveslaach** [comment]
If you are testing with [`nvim-lspconfig`](https://github.com/neovim/nvim-lspconfig) changing the CLI parameters is easy:

```lua
local lspconfig = require('lspconfig')
lspconfig.nushell.setup({
   cmd = { "nu", "--lsp", "-I", "… your path …", "-I", "… another path" },
   -- … more options …
})
```

> This may also assume that a persons env.nu is parsed when launching with nu --lsp which I'm not confident of.

It doesn't parse standard config files. Seems to me that it should in certain cases. Maybe, if there is no `rootUri` present. Let me check something.

**Comment by schrieveslaach** [comment]
@fdncred, have a look at #12454 if that solves your issue.

**Comment by fdncred** [comment]
> If you are testing with [nvim-lspconfig](https://github.com/neovim/nvim-lspconfig) changing the CLI parameters is easy:

This may work but I'm not confident with it because the `-I` parameter was never meant to be used multiple times when launching nushell. It's meant to be one string with separate folders separated by `\x1e`, iirc.

I'll test #12454 soon.

**Comment by schrieveslaach** [comment]
Providing `-I` multiple time was just a guess. Following should do the trick then.

```lua
local lspconfig = require('lspconfig')
lspconfig.nushell.setup({
   cmd = { "nu", "--lsp", "-I", "… your path …\x1e… another path" },
   -- … more options …
})
```

**Comment by tombh** [comment]
Regarding `nufmt`, the relevant part of the spec would be:
* [ ] [`textDocument/formatting`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_formatting)
